pkg/virt-handler/vm.go

```diff
@@ -1420,13 +1423,25 @@ func (d *VirtualMachineController) processVmUpdate(origVMI *v1.VirtualMachineIns
 				return err
 			}
 
+			res, err := d.podIsolationDetector.Detect(vmi)
+			if err != nil {
+				return fmt.Errorf("failed to detect isolation for migration target launcher pod: %v", err)
+			}
+
+			// configure network inside virt-launcher compute container
+			if err := res.DoNetNS(func() error {
```

Comment:
I think there's an optimization that needs to be taken into account here every time DoNetNS() is invoked.

If we tunnel through the code to where the OS thread is locked we'll see this code in vendor/github.com/containernetworking/plugins/pkg/ns/ns_linux.go
```
        var innerError error    
        go func() {                     
                defer wg.Done() 
                runtime.LockOSThread()
                innerError = containedCall(hostNS)                                                                  
        }()             
        wg.Wait()
```

Notice there that  `runtime.LockOSThread()` is locked, but never unlocked. I was curious what that meant, and found this comment on the `LockOSThread` function in /usr/lib/golang/src/runtime/proc.go

```
// If the calling goroutine exits without unlocking the thread,
// the thread will be terminated. 
```

This means every time we call DoNetNS() we're terminating a POSIX thread behind the scenes, which will probably result in the go runtime spinning up a replacement in the OS thread pool.

SO!

We need to make sure we only call DoNetNS if it is absolutely necessary. Caching when a successful completion of phase1 has occurred per a VMI might not be a bad idea. 

@roman @booxter ^ just something to think about.

Reply:
Good catch. I would suggest the following two things:

1) unlocking again:

```go
			// configure network inside virt-launcher compute container
			if err := res.DoNetNS(func() error {
                                defer runtime.UnlockOSThread()
				return network.SetupPodNetworkPhase1(vmi, res.Pid())
			}); err != nil {
				return fmt.Errorf("failed to configure vmi network for migration target: %v", err)
			}
```

then the switch is cheaper.

 2) https://github.com/kubevirt/kubevirt/pull/2837#discussion_r379416578

This would only lead to multiple switches if the startup fails, but then we have a retry backoff. In combination with unlocking this may already be good enough.

Reply:
And if we can unlock the thread probably also depends if we switch the netns back to the original one.

Reply:
Incredible catch. Thanks @davidvossel.